### PR TITLE
Strain and stress vectors have size 4 in 2D plane strain

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**Description**
Geomechanics application is refactored in a way that stress and strain vectors have size 4 in 2D plane strain.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_plane_strain_K0_law.cpp [hunk 1]
```diff
@@ -128,18 +134,9 @@ void LinearPlaneStrainK0Law::CalculatePK2Stress(const Vector& rStrainVector,
...
-    const double c0 = E / ((1.00 + NU)*(1 - 2 * NU));
-    const double c1 = (1.00 - NU)*c0;
-    const double c2 = c0 * NU;
-    const double c3 = (0.5 - NU)*c0;
-
-    rStressVector[INDEX_2D_PLANE_STRAIN_XX] = c1 * rStrainVector[INDEX_2D_PLANE_STRESS_XX] + c2 * rStrainVector[INDEX_2D_PLANE_STRESS_YY];
-    rStressVector[INDEX_2D_PLANE_STRAIN_YY] = c2 * rStrainVector[INDEX_2D_PLANE_STRESS_XX] + c1 * rStrainVector[INDEX_2D_PLANE_STRESS_YY];
-    rStressVector[INDEX_2D_PLANE_STRAIN_XY] = c3 * rStrainVector[INDEX_2D_PLANE_STRESS_XY];
-    rStressVector[INDEX_2D_PLANE_STRAIN_ZZ] = c2 * rStrainVector[INDEX_2D_PLANE_STRESS_XX] + c2 * rStrainVector[INDEX_2D_PLANE_STRESS_YY];
+    Matrix C;
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_plane_strain_K0_law.cpp:137]: Just mentioning that once  #8772 is merge you will be able to use bounded matrices in here, something that should make all these operations much more efficient.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp [hunk 2]
```diff
@@ -34,14 +34,14 @@ Element::Pointer UPwSmallStrainFICElement<TDim,TNumNodes>::Create(IndexType NewI
...
+void UPwSmallStrainFICElement<TDim,TNumNodes>::
+    Initialize(const ProcessInfo& rCurrentProcessInfo)
 {
     KRATOS_TRY
 
     UPwBaseElement<TDim,TNumNodes>::Initialize(rCurrentProcessInfo);
 
-    unsigned int VoigtSize = VOIGT_SIZE_3D;
-    if (TDim == 2) VoigtSize = VOIGT_SIZE_2D_PLANE_STRESS;
+    const SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp:44]:
```suggestion
    constexpr SizeType VoigtSize  = TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN;
```
This allows to create bounded matrices with `VoigtSize`. Indeed, you can specify this `constexpr` once in the begining of the class to use it al throughout the file.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp [hunk 3]
```diff
@@ -116,11 +117,8 @@ void UPwSmallStrainFICElement<TDim,TNumNodes>::
     }
     Matrix DtStressContainer(NumGPoints,TDim);
 
-    Vector StressVector;
-    if (TDim > 2)
-        StressVector.resize(VOIGT_SIZE_3D);
-    else
-        StressVector.resize(VOIGT_SIZE_2D_PLANE_STRESS);
+    const SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp:120]: Same `constexpr` comment.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp [hunk 4]
```diff
@@ -176,11 +175,8 @@ void UPwSmallStrainFICElement<TDim,TNumNodes>::
     //Containers for extrapolation variables
     Matrix DtStressContainer(NumGPoints,TDim);
 
-    Vector StressVector;
-    if (TDim > 2)
-        StressVector.resize(VOIGT_SIZE_3D);
-    else
-        StressVector.resize(VOIGT_SIZE_2D_PLANE_STRESS);
+    const SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp:178]: Same comment.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 5]
```diff
@@ -449,6 +454,8 @@ void UPwSmallStrainElement<TDim, TNumNodes>::
     array_1d<Vector, TNumNodes> NodalStressVector; //List with stresses at each node
     array_1d<Matrix, TNumNodes> NodalStressTensor;
 
+    const SizeType StressTensorSize = (TDim == N_DIM_3D ? STRESS_TENSOR_SIZE_3D : STRESS_TENSOR_SIZE_2D);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:457]:
```suggestion
    constexpr SizeType StressTensorSize = (TDim == N_DIM_3D ? STRESS_TENSOR_SIZE_3D : STRESS_TENSOR_SIZE_2D);
```

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 6]
```diff
@@ -860,13 +867,13 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
         GeometryType::ShapeFunctionsGradientsType DN_DXContainer(NumGPoints);
         Geom.ShapeFunctionsIntegrationPointsGradients(DN_DXContainer,this->GetIntegrationMethod());
 
-        unsigned int VoigtSize = VOIGT_SIZE_3D;
-        if (TDim == 2) VoigtSize = VOIGT_SIZE_2D_PLANE_STRESS;
-        Matrix B(VoigtSize,TNumNodes*TDim);
-        noalias(B) = ZeroMatrix(VoigtSize,TNumNodes*TDim);
+        const SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:870]:
```suggestion
        constexpr SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 7]
```diff
@@ -1232,8 +1239,7 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
     rVariables.detF = 1.0;
 
     //General Variables
-    unsigned int VoigtSize;
-    VoigtSize = (TDim > 2 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRESS);
+    const SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:1242]:
```suggestion
    constexpr SizeType VoigtSize  = ( TDim == 3 ? VOIGT_SIZE_3D : VOIGT_SIZE_2D_PLANE_STRAIN);
```

**Review by rubenzorrilla — APPROVED** [review]
Changes make sense to me. Only a minor comment about the use of `constexpr` for the Voigt sizes in the templated elements. Although right now this might have a not significance impact in the performance, it might be useful for the future (feel free to leave these comments for upcoming PRs).

**Comment by Vahid-Galavi** [comment]
> Changes make sense to me. Only a minor comment about the use of `constexpr` for the Voigt sizes in the templated elements. Although right now this might have a not significance impact in the performance, it might be useful for the future (feel free to leave these comments for upcoming PRs).

Thanks @rubenzorrilla! Good suggestions. I will consider them in the next PR.

**Pull request merged**